File: Administrador/App/screen/auth/register/register.viewmodel.ts
import { useState } from "react";
import useAuthentication from "../../../hooks/useAuthentication";

export const useRegisterViewModel = () => {
  const [nombrecompleto, setNombreCompleto] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [telefono, setTelefono] = useState("");
  const [error, setError] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  const { doRegister } = useAuthentication();

  const handleRegister = async () => {
    if (
      !email.trim() ||
      !password.trim() ||
      !nombrecompleto.trim() ||
      !telefono.trim()
    ) {
      setError("Todos los campos son requeridos");
      return;
    }

    setError("");
    setIsLoading(true);

    const loginData = {
      nombrecompleto: nombrecompleto.trim(),
      email: email.trim(),
      password: password,
      telefono: telefono.trim(),
    };

    const success = await doRegister(loginData);

    setIsLoading(false);

    if (!success) {
      setError("Error al iniciar sesión");
    }
  };

  return {
    nombrecompleto,
    setNombreCompleto,
    email,
    setEmail,
    password,
    setPassword,
    telefono,
    setTelefono,
    error,
    isLoading,
    handleRegister,
  };
};

File: Clientes/lib/screens/auth/register/register.viewmodel.dart
import 'package:flutter/material.dart';
import '../../../services/api_service.dart';
import '../../../models/auth_request.dart';

class RegisterViewModel extends ChangeNotifier {
  final ApiService _apiService = ApiService();

  bool _isLoading = false;
  String? _errorMessage;
  String? _successMessage;

  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;
  String? get successMessage => _successMessage;

  final TextEditingController emailController = TextEditingController();
  final TextEditingController nombreController = TextEditingController();
  final TextEditingController passwordController = TextEditingController();
  final TextEditingController telefonoController = TextEditingController();

  Future<bool> register() async {
    if (emailController.text.isEmpty ||
        nombreController.text.isEmpty ||
        passwordController.text.isEmpty ||
        telefonoController.text.isEmpty) {
      _errorMessage = 'Por favor complete todos los campos';
      notifyListeners();
      return false;
    }

    _isLoading = true;
    _errorMessage = null;
    _successMessage = null;
    notifyListeners();

    try {
      final request = RegisterRequest(
        email: emailController.text.trim(),
        nombrecompleto: nombreController.text.trim(),
        password: passwordController.text,
        telefono: telefonoController.text.trim(),
      );

      await _apiService.register(request);
      _isLoading = false;
      _successMessage = 'Registro exitoso. Ahora puedes iniciar sesión.';
      notifyListeners();
      return true;
    } catch (e) {
      _isLoading = false;
      _errorMessage = 'Error al registrarse: ${e.toString()}';
      notifyListeners();
      return false;
    }
  }

  @override
  void dispose() {
    emailController.dispose();
    nombreController.dispose();
    passwordController.dispose();
    telefonoController.dispose();
    super.dispose();
  }
}
